packages/lib/src/gfx/shaders/UberMaterial.js

```diff
@@ -197,187 +121,257 @@ UberMaterial.prototype.uberOptions = {
-  this.orthoCam = source.orthoCam;
-  this.prepassTransparancy = source.prepassTransparancy;
-  this.dashedLine = source.dashedLine;
-  this.thickLine = source.thickLine;
-  this.fogTransparent = source.fogTransparent;
-  this.normalsToGBuffer = source.normalsToGBuffer;
-  this.toonShading = source.toonShading;
+  precisionString() {
+    const { precision } = capabilities;
+    const str = `precision ${precision} float;\n`
+      + `precision ${precision} int;\n\n`;
+    return str;
+  }
 
-  this.uberOptions.copy(source.uberOptions);
+  // create copy of this material
+  copy(source) {
+    THREE.RawShaderMaterial.prototype.copy.call(this, source);
+
+    this.fragmentShader = source.fragmentShader;
+    this.vertexShader = source.vertexShader;
+
+    this.uniforms = THREE.UniformsUtils.clone(source.uniforms);
+    this.defines = { ...source.defines };
+    this.extensions = source.extensions;
+
+    this.fog = source.fog;
+    this.instancedPos = source.instancedPos;
+    this.instancedMatrix = source.instancedMatrix;
+    this.attrColor = source.attrColor;
+    this.attrColor2 = source.attrColor2;
+    this.attrAlphaColor = source.attrAlphaColor;
+    this.overrideColor = source.overrideColor;
+    this.sphereSprite = source.sphereSprite;
+    this.cylinderSprite = source.cylinderSprite;
+    this.zClip = source.zClip;
+    this.clipPlane = source.clipPlane;
+    this.fakeOpacity = source.fakeOpacity;
+    this.colorFromPos = source.colorFromPos;
+    this.shadowmap = source.shadowmap;
+    this.shadowmapType = source.shadowmapType;
+    this.colorFromDepth = source.colorFromDepth;
+    this.orthoCam = source.orthoCam;
+    this.prepassTransparancy = source.prepassTransparancy;
+    this.dashedLine = source.dashedLine;
+    this.thickLine = source.thickLine;
+    this.fogTransparent = source.fogTransparent;
+    this.normalsToGBuffer = source.normalsToGBuffer;
+    this.toonShading = source.toonShading;
 
-  return this;
-};
+    this.uberOptions.copy(source.uberOptions);
 
-// create copy of this material
-// its options are prototyped after this material's options
-UberMaterial.prototype.createInstance = function () {
-  const inst = new UberMaterial();
-  inst.copy(this);
-  inst.uberOptions = Object.create(this.uberOptions);
-  return inst;
-};
+    return this;
+  }
 
-UberMaterial.prototype.setValues = function (values) {
-  if (typeof values === 'undefined') {
-    return;
+  // its options are prototyped after this material's options
+  createInstance() {
+    const inst = new UberMaterial();
+    inst.copy(this);
+    inst.uberOptions = Object.create(this.uberOptions);
+    return inst;
   }
 
-  // set direct values
-  THREE.RawShaderMaterial.prototype.setValues.call(this, values);
+  setValues(values) {
+    if (typeof values === 'undefined') {
+      return;
+    }
 
-  const defines = {};
-  const extensions = {};
+    // set direct values
+    THREE.RawShaderMaterial.prototype.setValues.call(this, values);
```

Comment:
Can we use super.setValues(values) here?

Reply:
Yes we can